### PR TITLE
PATCH route added for partial updates

### PR DESCRIPTION
**Pull request by rombat**
related to https://axeptio.atlassian.net/browse/DEV-100
Once again, sorry about eslint/prettier polluting the diffs, real changes are `patch` and `patchDocument` functions
PATCH body request payload is an object compatible with the spec of the mongodb `$set` operator
Note: if the field is either an empty string, null or undefined, `$set` switches to `$unset`

### CODE REVIEW

File: services/docs/lib/services/document.js [hunk 1]
```diff
@@ -226,6 +226,41 @@ module.exports.setDocument = function(resource, filter, data, state, user) {
   });
 };
 
+module.exports.patchDocument = (resource, filter, data, state, user) => {
```
Comment (kevinrambaud) [inline services/docs/lib/services/document.js:229]: This function could be simplified if you used the promise signature of `resource.collection.findOneAndUpdate`, or even more, by using the async/await notation.

It's up to you if you want to change it.

File: services/docs/lib/services/document.js [hunk 2]
```diff
@@ -226,6 +226,41 @@ module.exports.setDocument = function(resource, filter, data, state, user) {
...
+            resolve({
+              id: filter._id,
+              state: state,
+              data: result.value.states[state].data
+            });
+          }
+        );
+      })
+      .catch(() => {
+        return reject(new Error('Validation Error'));
```
Comment (kevinrambaud) [inline services/docs/lib/services/document.js:259]: It sounds like this error could directly be thrown by the `builder.patch` function give the fact we're already swallowing the original error to throw this one.

File: services/docs/lib/modules/queryBuilder.js [hunk 3]
```diff
@@ -126,6 +126,32 @@ module.exports.update = function updateDoc(options) {
...
+              ? '$unset'
+              : '$set';
+          ops[operator][join('states', state.name, 'data', key)] = value;
+        }
+        if (Object.keys(ops.$unset).length === 0) {
+          delete ops.$unset;
+        }
+        return resolve(ops);
+      });
+  });
```
Comment (kevinrambaud) [inline services/docs/lib/modules/queryBuilder.js:152]: The content of this function could also be simplified with async/await notation.

File: services/docs/lib/services/document.js (hunk 1 above)
Comment (rombat) [reply services/docs/lib/services/document.js:229]: So far I've tried not to mess around too much with already existing functions in campsi-mono, but actually that can be a good exercise and improve readability so I'll do that 👍🏽

Comment (rombat) [reply services/docs/lib/services/document.js:229]: And actually it was not an already existing function anyway >_>

File: services/docs/lib/modules/queryBuilder.js [hunk 4]
```diff
@@ -126,6 +126,33 @@ module.exports.update = function updateDoc(options) {
   });
 };
 
+module.exports.patch = async options => {
+  const state = getStateFromOptions(options);
+  const validation = await validate(
```
Comment (kevinrambaud) [inline services/docs/lib/modules/queryBuilder.js:131]: It looks like the `validate` function either resolve true or reject (throw) an error. There is no false value that resolves.

I think the right thing to do to use the `validate` function as it is currently implemented would be to wrap the call to this function in a try/catch statement and throw the Validation Error you currently throw in the catch.

**Pull request merged**